aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/RetryHandler.java

```diff
@@ -0,0 +1,104 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.datastore.syncengine;
+
+import com.amplifyframework.datastore.utils.ErrorInspector;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import io.reactivex.rxjava3.core.Single;
+import io.reactivex.rxjava3.core.SingleEmitter;
+
+/**
+ * Class for retrying call on failure on a single.
+ */
+public class RetryHandler {
+    private final int maxExponentValue = 8;
+    private final int jitterFactorValue = 100;
+    private final int maxAttemptsValue = 3;
+    private final int maxExponent;
+    private final int jitterFactor;
+    private final int maxAttempts;
+
+    /**
+     * Constructor to inject constants for unit testing.
+     * @param maxExponent maxExponent backoff can go to.
+     * @param jitterFactor jitterFactor for backoff.
+     * @param maxAttempts max attempt for retrying.
+     */
+    public RetryHandler(int maxExponent,
+                        int jitterFactor,
+                        int maxAttempts) {
+
+        this.maxExponent = maxExponent;
+        this.jitterFactor = jitterFactor;
+        this.maxAttempts = maxAttempts;
+    }
+
+    /**
+     * Parameter less constructor.
+     */
+    public RetryHandler() {
+        maxExponent = maxExponentValue;
+        jitterFactor = jitterFactorValue;
+        maxAttempts = maxAttemptsValue;
+    }
+
+    /**
+     * retry.
+     * @param single single to be retried.
+     * @param skipExceptions exceptions which should not be retried.
+     * @param <T> The type for single.
+     * @return single of type T.
+     */
+    public <T> Single<T> retry(Single<T> single, List<Class<? extends Throwable>> skipExceptions) {
+        return Single.create(emitter -> call(single, emitter, 0L, maxAttempts, skipExceptions));
+
+    }
+
+    @SuppressWarnings("ResultOfMethodCallIgnored")
+    private <T> void call(
+            Single<T> single,
+            SingleEmitter<T> emitter,
+            Long delayInSeconds,
+            int attemptsLeft,
+            List<Class<? extends Throwable>> skipExceptions) {
+        single.delaySubscription(delayInSeconds, TimeUnit.SECONDS)
+                .subscribe(emitter::onSuccess, error -> {
+                    if (attemptsLeft == 0 || ErrorInspector.contains(error, skipExceptions)) {
+                        emitter.onError(error);
+                    } else {
+                        call(single, emitter, jitteredDelaySec(attemptsLeft),
+                                attemptsLeft - 1, skipExceptions);
+                    }
+                });
```

Comment:
How do you make sure that the flow was not canceled/disposed (which can lead to client app crashes)?

Reply:
Thanks for pointing this out. The single passed in this retry method is added to the [setDisposable on the emitter.](https://github.com/aws-amplify/amplify-android/pull/1414/files/a8f2d4d2be382af53d1045b95a25f57c531c09e6#:~:text=emitter.setDisposable(AmplifyDisposables.fromCancelable(cancelable))%3B)